summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/FutureQueue.scala

```diff
@@ -0,0 +1,202 @@
+   * ,at which finished futures are cleared.
+   * Clearing finished futures costs proportional to total number of outstanding
+   * futures, so we want to make sure we only clear when sufficient portion is
+   * finished.
+   */
+  val OutstandingFuturesDequeueRatio = 2
+
+  /**
+   * Wait for n futures to finish. Doesn't block, the returned future is satisfied
+   * once n futures have finished either successfully or unsuccessfully.
+   * If n is greater than number of futures in queue then we wait on all of them.
+   */
+  def waitN[A](fs: Iterable[Future[A]], n: Int): Future[Unit] = {
+    val waitOnCount = Math.min(fs.size, n)
+    if (waitOnCount <= 0) {
+      Future.Unit
+    } else {
+      val count = new AtomicInteger(waitOnCount)
+      val p = Promise[Unit]()
+      fs.foreach { f =>
+        f.ensure {
+          // Note that since we are only decrementing we can cross 0 only
+          // once (unless we decrement more than 2^32 times).
+          if (count.decrementAndGet() == 0) {
+            p.setValue(())
+          }
+        }
+      }
+      p
+    }
+  }
+}
+
+/**
+ * Use this class to do lazy continuations.  This is especially useful
+ * when using a single thread that initiates Futures and must also handle
+ * the continuation.
+ *
+ * type S: state associated with each Future
+ * type T: the result type of each Future
+ *
+ * Minimally, this queue accepts a state value along with a Future.  Once
+ * the Future completes, the result can be retrieved, along with the state,
+ * via the dequeue method.
+ *
+ * To support batching, inputs can also be TraversableOnce[(S, Future[T])]
+ *
+ * To support additional asynchronous behavior, a state value along with a
+ * Future[TraversableOnce[(S, Future[T])]] can be inserted.  On failure, the
+ * outer state value is returned with the Failure.  On success, the inner
+ * items are queued up to be inserted when they complete.
+ */
+private[summingbird] class FutureQueue[S, T](
+    maxWaitingFutures: MaxWaitingFutures,
+    maxWaitingTime: MaxFutureWaitTime,
+    maxEmitPerExec: MaxEmitPerExecute) {
+  @transient protected lazy val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  // Track the futures that may be outstanding.  We may need to wait on some
+  // of them in case the actual outstanding count gets too high.
+  private lazy val outstandingFutures = Queue.linkedNonBlocking[Future[Unit]]
+  // Track the number of actually outstanding futures
+  private[executor] lazy val numPendingOutstandingFutures = new AtomicInteger(0)
+  // When futures complete, they deposit their results here.
+  private lazy val responses = Queue.linkedNonBlocking[(S, Try[T])]
+
+  private val tryBijection = UtilBijections.twitter2ScalaTry[T]
+
+  /**
+   * Add a Future with state onto the queue.  This can eventually be
+   * dequeued as (S, Try[T]) when the Future completes.
+   *
+   * Returns true if the Future is not yet complete.  When false, the
```

Comment:
Fair point.  I was anyway a little worried that folks would confuse this with "item was queued or not" semantics.  Making this a `Unit` method simplifies that; I'll make this a private internal method and making a public wrapper for it.